File: crates/yokan/demo/loading.py
# /// script
# requires-python = ">=3.14"
# ///
"""progress: `value` stays the only required prop — `width`/`height`
size the track, `label` draws a dim line above it, and
`indeterminate` ignores `value` and sweeps a segment instead, for
work with no known length.

Develop:  uv run demo/loading.py
Ship:     python3 yokan_gate.py gate demo/loading.py --script "click:step,click:step,dump,click:busy"
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from yokan import button, column, progress, row, run, State, text  # noqa: E402


ratio: State[float] = State(0.25)
busy: State[bool] = State(False)


def step():
    if ratio() >= 1.0:
        ratio.set(0.0)
    else:
        ratio.set(ratio() + 0.25)


def toggle_busy():
    busy.set(not busy())


def view():
    return column(
        text(f"ratio: {ratio()}"),
        progress(ratio(), label="Uploading"),
        progress(ratio(), width=240, height=6),
        progress(ratio(), indeterminate=busy()),
        row(
            button("step", on_click=step),
            button("busy", on_click=toggle_busy),
            spacing=8,
        ),
        spacing=12,
        padding=16,
    )


if __name__ == "__main__":
    run(view, title="loading")

File: gomamochi/demo/loading.go
// The bar that fills, in its three forms: with a caption above it, at a
// size the app chose, and sweeping for work with no known length.
package main

import (
	"strconv"
	"strings"

	. "github.com/i2y/yokan/gomamochi"
)

type Loading struct {
	ratio float64
	busy  bool
}

func (l *Loading) step() {
	if l.ratio >= 1 {
		l.ratio = 0
	} else {
		l.ratio += 0.25
	}
}

// A number the way the other languages print one: a whole number
// still shows its ".0".
func num(v float64) string {
	s := strconv.FormatFloat(v, 'f', -1, 64)
	if !strings.ContainsAny(s, ".e") {
		s += ".0"
	}
	return s
}

func (l *Loading) View() Element {
	return Column(
		Text("ratio: "+num(l.ratio)),
		Progress(l.ratio).Label("Uploading"),
		Progress(l.ratio).Width(240).Height(6),
		Progress(l.ratio).Indeterminate(l.busy),
		Row(
			Button("step").OnClick(func() { l.step() }),
			Button("busy").OnClick(func() { l.busy = !l.busy }),
		).Spacing(8),
	).Spacing(12).Padding(16)
}

func main() {
	Run(&Loading{ratio: 0.25}, Title("loading"))
}
